controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -275,6 +282,184 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
+                           } else if (cause instanceof TimeoutException) {
+                               throw new CompletionException(cause);
+                           } else {
+                               log.warn(requestId, "Remove subscriber from stream failed due to ", cause);
+                               return RemoveSubscriberStatus.Status.FAILURE;
+                           }
+                       });
+           }), e -> Exceptions.unwrap(e) instanceof RetryableException, SUBSCRIBER_OPERATION_RETRIES, executor);
+    }
+
+
+    /**
+     * Get list of  subscribers for Stream.
+     * Needed only for Consumption based retention.
+     * @param scope      scope.
+     * @param stream     stream name.
+     * @param contextOpt optional context
+     * @return update status.
+     */
+    public CompletableFuture<List<String>> getSubscribersForStream(String scope, String stream, OperationContext contextOpt) {
+        final OperationContext context = contextOpt == null ? streamMetadataStore.createContext(scope, stream) : contextOpt;
+        final long requestId = requestTracker.getRequestIdFor("getSubscribersForStream", scope, stream);
+        final List<String> emptySubscribersList = Collections.emptyList();
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.checkStreamExists(scope, stream)
+                .thenCompose(exists -> {
+                  if (!exists) {
+                    // 1. if Stream does not exist return empty list
+                    return CompletableFuture.completedFuture(emptySubscribersList);
+                  }
+                  // 2. get subscribers data
+                  return Futures.exceptionallyExpecting(streamMetadataStore.getAllSubscribers(scope, stream, context, executor),
+                    e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, null)
+                    .thenCompose(subscribersData -> {
+                        //3. if SubscribersRecord does not exist return empty list
+                        if (subscribersData == null) {
+                            return CompletableFuture.completedFuture(emptySubscribersList);
+                        }
+                        return CompletableFuture.completedFuture(subscribersData.keySet().stream().collect(Collectors.toList()));
+                    }).exceptionally(ex -> {
+                              log.warn(requestId, "Exception thrown in trying to get list of Stream subscribers. {}",
+                                      ex.getMessage());
+                              Throwable cause = Exceptions.unwrap(ex);
+                              if (cause instanceof TimeoutException) {
+                                  throw new CompletionException(cause);
+                              } else {
+                                  log.warn(requestId, "getSubscribersForStream failed due to {}", cause);
+                                  return emptySubscribersList;
```

Comment:
why return empty subscriber list when there is an exception?

Reply:
The DataNotFoundException here indicates there is no subscriber data present for this Stream, which can happen if users query subscriber data for a Stream that does not have "subscribers". So this is not really an exceptional case and hence returning empty list here.

Reply:
but can empty list also be returned if the stream did not exist?